dcos_installer/upgrade.py

```diff
@@ -0,0 +1,128 @@
+'''
+Generating node upgrade script
+'''
+
+import subprocess
+
+import gen.build_deploy.util as util
+import gen.template
+from dcos_installer.constants import SERVE_DIR
+from pkgpanda.util import hash_str, write_string
+
+
+node_upgrade_template = """
+#!/bin/bash
+#
+# BASH script to upgrade DC/OS on a node
+#
+# Metadata:
+#   dcos image commit: {{ dcos_image_commit }}
+#   generation date: {{ generation_date }}
+# Copyright 2017 Mesosphere, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit -o nounset -o pipefail
+
+source /opt/mesosphere/environment.export
+
+# Check if this is a terminal, and if colors are supported, set some basic
+# colors for outputs
+if [ -t 1 ]; then
+    colors_supported=$(tput colors)
+    if [[ $colors_supported -ge 8 ]]; then
+    RED='\e[1;31m'
+    BOLD='\e[1m'
+    NORMAL='\e[0m'
+    fi
+fi
+
+if [[ $EUID -ne 0 ]]; then
+    echo "This script must be run as root" 1>&2
+    exit 1
+fi
+
+
+# check for version of dc/os upgrading from
+version=`grep "version" /opt/mesosphere/etc/dcos-version.json | cut -d '"' -f 4`
+if [ $version != {{ dcos_version }} ]; then
+   echo "Version specified is not correct. Version expected ${version}"
+   exit 1
+fi
+
+echo "Fetching packages"
+
+pkgpanda fetch --repository-url={{ bootstrap_url }} {{ cluster_packages }}
+
+echo "Activating packages"
+
+pkgpanda activate --no-block {{ cluster_packages }}
+
+# check if we are on a master node
+if [ -f /etc/mesosphere/roles/master ]; then
+   # run exhibitor migration script here
+   until dcos-shell dcos-exhibitor-migrate-perform
+   do
+        status=$?
+        case $status in
+             1) echo "Waiting for exhibitor endpoint"
+                sleep 10
+                ;;
+             2) echo "Could not read from exhibitor"
+                echo "Contact Support"
+                exit
+                ;;
+             4) echo "Rolling update in progress"
+                exit
+                ;;
+             8) echo "At least one config value does not have the expected pre migration value,
+                      \ and automatic migration can not take place"
+                echo "Contact Support"
+                exit
+                ;;
+             16) echo "Attempting to start the rolling update failed due to a non 200 response from exhibitor"
+                 echo "Contact Support"
+                 exit
+                 ;;
+        esac
+   done
+fi
+
+echo "Node upgrade complete"
```

Comment:
This message shouldn't be here, the upgrade isn't complete until it's validated, which this script doesn't do. The design doc also has us printing out the version before we start the upgrade. Please re-read the design doc and adjust this script's output to match it.